File: lbr_demos/lbr_demos_cpp/src/joint_sine_overlay.cpp
#include <functional>
#include <math.h>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"

// include fri for session state
#include "friClientIf.h"

// include lbr_fri_idl
#include "lbr_fri_idl/msg/lbr_joint_position_command.hpp"
#include "lbr_fri_idl/msg/lbr_state.hpp"
#include "lbr_fri_ros2/utils.hpp"

class JointSineOverlay {
  static constexpr double amplitude_ = 0.04; // rad
  static constexpr double frequency_ = 0.25; // Hz

public:
  JointSineOverlay(const rclcpp::Node::SharedPtr node) : node_(node), phase_(0.) {
    // create publisher to command/joint_position
    lbr_joint_position_command_pub_ =
        node_->create_publisher<lbr_fri_idl::msg::LBRJointPositionCommand>("command/joint_position",
                                                                           1);

    // create subscription to state
    lbr_state_sub_ = node_->create_subscription<lbr_fri_idl::msg::LBRState>(
        "state", 1, std::bind(&JointSineOverlay::on_lbr_state_, this, std::placeholders::_1));

    // get control rate from controller_manager
    auto update_rate =
        lbr_fri_ros2::retrieve_paramter(node_, "controller_manager", "update_rate").as_int();
    dt_ = 1.0 / static_cast<double>(update_rate);
  };

protected:
  void on_lbr_state_(const lbr_fri_idl::msg::LBRState::SharedPtr lbr_state) {
    if (!lbr_state_init_) {
      lbr_state_ = *lbr_state;
      lbr_state_init_ = true;
    }
    lbr_joint_position_command_.joint_position = lbr_state_.measured_joint_position;

    if (lbr_state->session_state == KUKA::FRI::COMMANDING_ACTIVE) {
      // overlay sine wave on 4th joint
      lbr_joint_position_command_.joint_position[3] += amplitude_ * sin(phase_);
      phase_ += 2 * M_PI * frequency_ * dt_;

      lbr_joint_position_command_pub_->publish(lbr_joint_position_command_);
    } else {
      // reset phase
      phase_ = 0.;
    }
  };

protected:
  rclcpp::Node::SharedPtr node_;
  double dt_;
  double phase_;
  rclcpp::Publisher<lbr_fri_idl::msg::LBRJointPositionCommand>::SharedPtr
      lbr_joint_position_command_pub_;
  rclcpp::Subscription<lbr_fri_idl::msg::LBRState>::SharedPtr lbr_state_sub_;
  bool lbr_state_init_ = false;
  lbr_fri_idl::msg::LBRState lbr_state_;
  lbr_fri_idl::msg::LBRJointPositionCommand lbr_joint_position_command_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<rclcpp::Node>("joint_sine_overlay");
  JointSineOverlay joint_sine_overlay(node);
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
};

File: lbr_demos/lbr_demos_py/lbr_demos_py/joint_sine_overlay.py
import math
from copy import deepcopy

import rclpy
from rcl_interfaces.srv import GetParameters
from rclpy.node import Node

# import lbr_fri_idl
from lbr_fri_idl.msg import LBRJointPositionCommand, LBRState


class JointSineOverlayNode(Node):
    def __init__(self, node_name: str) -> None:
        super().__init__(node_name)

        self._amplitude = 0.04  # rad
        self._frequency = 0.25  # Hz
        self._phase = 0.0
        self._lbr_joint_position_command = LBRJointPositionCommand()

        # create publisher to command/joint_position
        self._lbr_joint_position_command_pub = self.create_publisher(
            LBRJointPositionCommand,
            "command/joint_position",
            1,
        )

        # create subscription to state
        self._lbr_state = None
        self._lbr_state_sub_ = self.create_subscription(
            LBRState, "state", self._on_lbr_state, 1
        )

        # get control rate from controller_manager
        self._dt = None
        self._retrieve_update_rate()

    def _on_lbr_state(self, lbr_state: LBRState) -> None:
        if self._dt is None:
            return
        if self._lbr_state is None:
            self._lbr_state = lbr_state
        self._lbr_joint_position_command.joint_position = deepcopy(
            self._lbr_state.measured_joint_position
        )

        if lbr_state.session_state == 4:  # KUKA::FRI::COMMANDING_ACTIVE == 4
            # overlay sine wave on 4th joint
            self._lbr_joint_position_command.joint_position[
                3
            ] += self._amplitude * math.sin(self._phase)
            self._phase += 2 * math.pi * self._frequency * self._dt

            self._lbr_joint_position_command_pub.publish(
                self._lbr_joint_position_command
            )
        else:
            # reset phase
            self._phase = 0.0

    def _retrieve_update_rate(self) -> float:
        paramter_client = self.create_client(
            GetParameters, "controller_manager/get_parameters"
        )
        paramter_name = "update_rate"
        while not paramter_client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                raise RuntimeError("Interrupted while waiting for service.")
            self.get_logger().info(f"Waiting for {paramter_client.srv_name}...")
        future = paramter_client.call_async(
            GetParameters.Request(names=[paramter_name])
        )
        rclpy.spin_until_future_complete(self, future)
        if future.result() is None:
            raise RuntimeError(f"Failed to get parameter '{paramter_name}'.")
        update_rate = future.result().values[0].integer_value
        self.get_logger().info(f"{paramter_name}: {update_rate} Hz")
        self._dt = 1.0 / float(update_rate)


def main(args: list = None) -> None:
    rclpy.init(args=args)
    rclpy.spin(JointSineOverlayNode("joint_sine_overlay_node"))
    rclpy.shutdown()
